pkg/util/stringencoding/string_encoding.go

```diff
@@ -109,11 +108,25 @@ func EncodeEscapedChar(
 			// Escape non-printable characters.
 			buf.Write(HexMap[currentByte])
 		}
-	} else if ln == 2 {
-		// For multi-byte runes, print them based on their width.
-		fmt.Fprintf(buf, `\u%04X`, currentRune)
 	} else {
-		fmt.Fprintf(buf, `\U%08X`, currentRune)
+		writeMultibyteRuneAsHex(buf, currentRune, ln)
+	}
+}
+
+const uppercaseHex = `0123456789ABCDEF`
+
+// writeMultibyteRuneAsHex is equivalent to either
+// fmt.FPrintf(`\u%04X`) or fmt.FPrintf(`\U%08X`).
+// We can't quite just use strconv since we need uppercase hex.
+func writeMultibyteRuneAsHex(buf *bytes.Buffer, r rune, ln int) {
+	if ln == 2 {
+		buf.WriteString(`\u0000`)
+	} else {
+		buf.WriteString(`\U00000000`)
+	}
+	for i := 1; r > 0; r /= 16 {
```

Comment:
Probably better to do something like hex.Encode does vs mod/div.
```
dst[j] = hextable[v>>4]
dst[j+1] = hextable[v&0x0f]
```

Reply:
I really hope the compiler just does this optimization on its own but I guess it can't hurt to just write it that way, thanks.